modules/programs/ncmpcpp.nix

```
@@ -0,0 +1,136 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) types mkIf mkEnableOption mkOption;
+
+  cfg = config.programs.ncmpcpp;
+
+  mpdCfg = config.services.mpd;
+
+  renderSettings = settings:
+    lib.concatStringsSep "\n" (lib.mapAttrsToList renderSetting settings);
+
+  renderSetting = name: value: "${name}=${renderValue value}";
+
+  renderValue = option:
+    {
+      int = toString option;
+      bool = if option then "yes" else "no";
+      string = option;
+    }.${builtins.typeOf option};
+
+  renderBindings = bindings:
+    lib.concatStringsSep "\n" (map renderBinding bindings);
+
+  renderBinding = { key, command }:
+    lib.concatStringsSep "\n  "
+    ([ ''def_key "${key}"'' ] ++ maybeWrapList command);
+
+  maybeWrapList = xs: if lib.isList xs then xs else [ xs ];
+
+  valueType = with types; oneOf [ bool int str ];
+
+  bindingType = types.submodule ({ name, config, ... }: {
+    options = {
+      key = mkOption {
+        type = types.str;
+        description = "Key to bind.";
+        example = "j";
+      };
+
+      command = mkOption {
+        type = with types; either str (listOf str);
+        description = "Command or sequence of commands to be executed.";
+        example = "scroll_down";
+      };
+    };
+  });
+
+in {
+  meta.maintainers = with lib.maintainers; [ olmokramer ];
+
+  options.programs.ncmpcpp = {
+    enable =
+      mkEnableOption "ncmpcpp - An ncurses Music Player Daemon (MPD) client";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.ncmpcpp;
+      defaultText = "pkgs.ncmpcpp";
+      description = "Package providing the <code>ncmpcpp</code> command.";
+      example = "pkgs.ncmpcpp.override { ... }";
+    };
+
+    mpdMusicDir = mkOption {
+      type = types.nullOr types.path;
+      default = if mpdCfg.enable then mpdCfg.musicDirectory else null;
```

I suspect this also should verify that the target platform is Linux based since it's more of a historic accident that the mpd module is available on Darwin. Ideally we'll remove the module from Darwin builds in the future and when that happens this code would break. Perhaps something like
```suggestion
      default = if pkgs.stdenv.hostPlatform.isLinux && mpdCfg.enable then
        mpdCfg.musicDirectory
      else
        null;
```
would work.

---

Yeah I also wondered if this would work on MacOS, but I figured that since the MPD service module is available on MacOS that it would simply work. If the service should realistically only be available on Linux this change seems sensible to me, since a google search for "ncmpcpp mac osx" gives me some results that implicate that ncmpcpp should be working fine on MacOS, so disabling the entire module on MacOS wouldn't make sense.